### PR TITLE
[tools][builder][gwf] Rewrite gwf translator from py to cpp

### PR DESCRIPTION
**Pull request by IRomanchuk06**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [ ] Update documentation

### CODE REVIEW

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
+#include "gwf_parser.hpp"
+
+std::string GwfParser::XmlCharToString(std::unique_ptr<xmlChar, XmlCharDeleter> const & ptr)
+{
+  return ptr ? std::string(reinterpret_cast<char const *>(ptr.get())) : "";
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:5]: Are there any ways to convert xmlChar array to std::string? Using cast in this code can be dangerous

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+std::unique_ptr<xmlChar, XmlCharDeleter> GwfParser::GetXmlProp(xmlNodePtr node, char const * propName)
+{
+  return std::unique_ptr<xmlChar, XmlCharDeleter>(xmlGetProp(node, reinterpret_cast<xmlChar const *>(propName)));
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:10]: Think about it also

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  }
+  else if (tag == "pair" || tag == "arc")
+  {
+    ProcessPair(child, attributes);
+  }
+  else if (tag == "bus")
+  {
+    ProcessBus(child, attributes);
+  }
+  else if (tag == "contour")
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:28]: This case don't contain any logic. Why?

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      auto const mimeTypeAttr = GetXmlProp(contentChild, "mime_type");
+      auto const fileNameAttr = GetXmlProp(contentChild, "file_name");
+      auto const contentData = std::unique_ptr<xmlChar, XmlCharDeleter>(xmlNodeGetContent(contentChild));
+
+      attributes["content_type"] = XmlCharToString(contentTypeAttr);
+      attributes["mime_type"] = XmlCharToString(mimeTypeAttr);
+      attributes["file_name"] = XmlCharToString(fileNameAttr);
+
+      std::string const & contentType = attributes["content_type"];
+      if (!contentType.empty() && std::stoi(contentType) < 4)
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:54]: If contentType can't be convert to int then std::stoi will cause exception

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      {
+        attributes["content_data"] = XmlCharToString(contentData);
+      }
+      else if (contentType == "4")
+      {
+        attributes["content_data"] = ScBase64::Decode(XmlCharToString(contentData));
+      }
+      else
+      {
+        throw std::runtime_error("Content type is not supported: " + contentType);
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:64]: We use sc-exceptions

Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:64]:
```suggestion
        SC_THROW_EXCEPTION(utils::ParseError, "Content type is not supported: " + contentType);
```

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  {
+    if (contentChild->type == XML_ELEMENT_NODE
+        && xmlStrEqual(contentChild->name, reinterpret_cast<xmlChar const *>("content")))
+    {
+      auto const contentTypeAttr = GetXmlProp(contentChild, "type");
+      auto const mimeTypeAttr = GetXmlProp(contentChild, "mime_type");
+      auto const fileNameAttr = GetXmlProp(contentChild, "file_name");
+      auto const contentData = std::unique_ptr<xmlChar, XmlCharDeleter>(xmlNodeGetContent(contentChild));
+
+      attributes["content_type"] = XmlCharToString(contentTypeAttr);
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:49]: Let implement hierarchy of sc.g-elements and incapsulate this fields in correspondent classes. You had better to do this, it can help to support this code in the future

File: sc-tools/sc-builder/src/gwf_parser.cpp (hunk 5 above)
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/gwf_parser.cpp:64]: Review the entire code

File: sc-tools/sc-builder/src/gwf_parser.cpp (hunk 6 above)
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/gwf_parser.cpp:49]: I propose this:

```
SCgElement
-> SCgNode
     -> SCgLink
     -> SCgContour
     -> SCgBus
-> SCgConnector
     -> SCgEdge
     -> SCgArc
```

File: sc-tools/sc-builder/src/gwf_parser.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    {
+      throw std::runtime_error("Failed to open XML xmlTree.");
+    }
+
+    xmlNodePtr rootElement = xmlDocGetRootElement(xmlTree.get());
+    xmlNodePtr staticSector = nullptr;
+
+    for (xmlNodePtr node = rootElement->children; node != nullptr; node = node->next)
+    {
+      if (xmlStrEqual(node->name, reinterpret_cast<xmlChar const *>("staticSector")))
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_parser.cpp:141]: I think that there is more good way to convert std::string to array of xmlChar

File: sc-tools/sc-builder/src/gwf_translator.cpp [hunk 8]
```diff
@@ -10,60 +10,111 @@
...
+  const std::string scsStr = writer.Write(elements, filePath);
+
+  return scsStr;
 }
 
-bool GWFTranslator::ErrorsExist(std::string const & path) const
+std::string GWFTranslator::WriteStringToFile(std::string const & scsStr, std::string const & filePath)
 {
-  return std::filesystem::is_regular_file(path) && std::filesystem::file_size(path) > 0;
+  const std::string scsSource = filePath + ".scs";
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_translator.cpp:41]: We use const-after-type style for all variables, functions and all other objects in C++ and C

File: sc-tools/sc-builder/src/gwf_translator.hpp [hunk 9]
```diff
@@ -10,6 +10,9 @@
 #include "scs_translator.hpp"
 #include "builder_defines.hpp"
 
+#include <vector>
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/gwf_translator.hpp:13]: They must be before all includes

File: sc-tools/sc-builder/src/sc_scs_writer.cpp [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+}
+
+std::string Buffer::GetValue() const
+{
+  return value;
+}
+
+// UTILS CLASS
+
+std::string Utils::m_ReplaceAll(std::string const & str, std::string const & from, std::string const & to)
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/sc_scs_writer.cpp:37]: And we have own utils for std::string operations. Use `#include <sc-memory/sc_utils.hpp>`

File: sc-tools/sc-builder/src/sc_scs_writer.cpp [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+
+void ScsWriter::ProcessElementsList(
+    std::vector<std::unordered_map<std::string, std::string>> const & elementsList,
+    Buffer & buffer,
+    std::string const & parent,
+    std::size_t nestedLevel)
+{
+  try
+  {
+    std::vector<std::unordered_map<std::string, std::string>> contoursList;
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/sc_scs_writer.cpp:107]: You can select interfaces in base abstract class. And implement it in all your derived classes

File: sc-tools/sc-builder/src/sc_scs_writer.cpp [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+
+  auto FindElementById = [&](std::string const & id) -> std::unordered_map<std::string, std::string>
+  {
+    auto foundElement = std::find_if(
+        elementsList.begin(),
+        elementsList.end(),
+        [&](const auto & element)
+        {
+          auto it = element.find("id");
+          return (it != element.end() && it->second == id);
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/sc_scs_writer.cpp:228]: Review everywhere, please

File: sc-tools/sc-builder/src/scg_to_scs_el.cpp [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+      {"EdgeTypes", &m_edgeTypes},
+      {"BackwardEdgeTypes", &m_backwardEdgeTypes},
+      {"UnsupportedEdgeTypes", &m_unsupportedEdgeTypes}};
+
+  auto it = dictMap.find(dict);
+  if (it != dictMap.end())
+  {
+    return m_FindValue(*(it->second), scgElement);
+  }
+  return "";
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/scg_to_scs_el.cpp:246]:
```suggestion
  return "";
```

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 14]
```diff
@@ -8,30 +8,119 @@
...
 #include "builder_test.hpp"
 
+#include <filesystem>
+
 #include "sc-memory/sc_utils.hpp"
 #include "../../src/gwf_translator.hpp"
 
-#define TEST_CORRECT_GWF SC_BUILDER_KB "/test-templates/test_template_1.gwf"
-#define TEST_INCORRECT_GWF SC_BUILDER_TEST_REPOS "/kb/example.gwf"
+#define BASE_TEST_PATH "/home/iromanchuk/sc-machine/sc-tools/sc-builder/tests/kb/tests-gwf-to-scs/"
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:16]: Let I show you how your code can generate this strings automatically

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 15]
```diff
@@ -8,30 +8,119 @@
...
+  GWFTranslator translator(*m_ctx);
+
+  const std::string filePath = BASE_TEST_PATH "empty_contour.gwf";
+
+  const std::string gwfStr = translator.XmlFileToString(filePath);
+
+  const std::string scsStr = translator.GwfToScs(gwfStr, filePath);
+
+  EXPECT_TRUE(compareStringWithFileContent(scsStr, filePath + ".scs"));
+}
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:126]:
```suggestion
}

```

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sc-tools/sc-builder/src/scg_to_scs_el.cpp [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+
+std::string ScgToScsElement::GetElement(std::string const & scgElement, std::string const & dict)
+{
+  static const std::unordered_map<std::string, std::unordered_map<std::string, std::string> const *> dictMap = {
+      {"NodeTypeSets", &m_nodeTypeSets},
+      {"BackwardNodeTypes", &m_backwardNodeTypes},
+      {"UnsupportedNodeTypeSets", &m_unsupportedNodeTypeSets},
+      {"EdgeTypes", &m_edgeTypes},
+      {"BackwardEdgeTypes", &m_backwardEdgeTypes},
+      {"UnsupportedEdgeTypes", &m_unsupportedEdgeTypes}};
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/src/scg_to_scs_el.cpp:239]: to reduce possibility of error I'd create methods like getNodeType(), getBackwardNodeType() etc.
in that case user won't be using string literals to get required dictionary

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 17]
```diff
@@ -8,30 +8,142 @@
 
 #include "builder_test.hpp"
 
+#include <filesystem>
+
 #include "sc-memory/sc_utils.hpp"
 #include "../../src/gwf_translator.hpp"
+#include "../../src/sc_scs_tree.hpp"
+
+#define BASE_TEST_PATH "/home/iromanchuk/sc-machine/sc-tools/sc-builder/tests/kb/tests-gwf-to-scs/"
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:17]: I guess you can use `SC_BUILDER_KB "/tests-gwf-to-scs/"` instead of complete path

File: sc-tools/sc-builder/tests/units/sc_scs_tree.cpp [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+        isContentMultiLine = false;
+        multiLineContent.clear();
+
+        continue;
+      }
+
+      multiLineContent += "\n" + line;
+      continue;
+    }
+    else if (flag)
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/sc_scs_tree.cpp:89]: isContentMultiLine is understandable flag name, this flag's name is not

File: sc-tools/sc-builder/tests/units/sc_scs_tree.cpp [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+
+      if (Check(CONNECTOR_BEGIN, line))
+      {
+        size_t edgeBeginPos = line.find(CONNECTOR_BEGIN) + CONNECTOR_BEGIN.length();
+
+        if (Check(CONNECTOR_END, line))
+        {
+          size_t edgeEndPos = line.find(CONNECTOR_END);
+
+          std::string content = line.substr(edgeBeginPos, edgeEndPos - edgeBeginPos);
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/sc_scs_tree.cpp:117]: 'connector' should be used instead of edge

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 20]
```diff
@@ -4,33 +4,209 @@
...
+  std::string const & testGWFText = translator.GetXMLFileContent(testGWFFilePath);
+  std::string const & testSCsText = translator.TranslateGWFToSCs(testGWFText, BASE_TEST_PATH);
+
+  std::string const & correctSCsText = ReadFileToString(correctSCsFilePath);
+
+  SCsTree testSCsTree;
+  testSCsTree.Parse(testSCsText);
+  SCsTree correctSCsTree;
+  correctSCsTree.Parse(correctSCsText);
+  return testSCsTree.Compare(correctSCsTree)->empty();
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:71]: this Compare returns list with differences so for test it'll be helpful that if it is not empty then those differences are printed into some log

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 21]
```diff
@@ -4,33 +4,209 @@
...
-  translateParams.m_fileName = TEST_CORRECT_GWF;
-  translateParams.m_autoFormatInfo = true;
-  translateParams.m_outputStructure = ScAddr::Empty;
-  EXPECT_TRUE(translator.Translate(translateParams));
+  std::string const filePath = BASE_TEST_PATH "empty_file.gwf";
+  EXPECT_THROW(translator.GetXMLFileContent(filePath), utils::ExceptionParseError);
 }
 
-TEST_F(ScBuilderTest, IncorrectGWF)
+TEST_F(GWFTranslatorTest, EmptyStatic)
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:90]: also add test FileWithoutStatic to see what'll happen if file is not empty and it has no static sector

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 22]
```diff
@@ -4,33 +4,209 @@
...
   GWFTranslator translator(*m_ctx);
 
-  Translator::Params translateParams;
-  translateParams.m_fileName = TEST_INCORRECT_GWF;
-  translateParams.m_autoFormatInfo = true;
-  translateParams.m_outputStructure = ScAddr::Empty;
-  EXPECT_THROW(translator.Translate(translateParams), utils::ExceptionParseError);
+  std::string const & filePath = BASE_TEST_PATH "empty_static.gwf";
+  std::string const & gwfText = translator.GetXMLFileContent(filePath);
+  EXPECT_THROW(translator.TranslateGWFToSCs(gwfText, BASE_TEST_PATH), utils::ExceptionParseError);
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:96]:
```suggestion
  EXPECT_THROW(translator.TranslateGWFToSCs(gwfText, filePath), utils::ExceptionParseError);
```
shouldn't filePath be passed here? check other tests as well

File: sc-tools/sc-builder/tests/units/test_gwf_translator.cpp [hunk 23]
```diff
@@ -4,33 +4,209 @@
...
+  std::string const & testSCsText = translator.TranslateGWFToSCs(testGWFText, BASE_TEST_PATH);
+
+  std::string const & correctSCsText = ReadFileToString(correctSCsFilePath);
+
+  SCsTree testSCsTree;
+  testSCsTree.Parse(testSCsText);
+  SCsTree correctSCsTree;
+  correctSCsTree.Parse(correctSCsText);
+  EXPECT_FALSE(testSCsTree.Compare(correctSCsTree)->empty());
+  EXPECT_FALSE(correctSCsTree.Compare(testSCsTree)->empty());
```
Comment (kilativ-dotcom) [inline sc-tools/sc-builder/tests/units/test_gwf_translator.cpp:191]: I don't think that these two checks are helpful, they are just saying that trees are different but without specifying what those differences should be

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

File: sc-tools/sc-builder/src/sc_scs_element.cpp [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+
+void SCsLink::ConvertFromSCgElement(SCgElementPtr const & scgElement)
+{
+  auto link = std::dynamic_pointer_cast<SCgLink>(scgElement);
+  if (link == nullptr)
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidType, "SCsLink::ConvertFromSCgElement: Invalid SCgElement passed to SCsLink.");
+
+  ContentType contentType = static_cast<ContentType>(std::stoi(link->GetContentType()));
+  m_type = link->GetType();
```
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-tools/sc-builder/src/sc_scs_element.cpp:92]: type is not printed to scs and just used to determine variability

**Review by NikitaZotov — APPROVED** [review]

**Review by kilativ-dotcom — APPROVED** [review]
